### PR TITLE
Die with dignity on the network layer

### PR DESCRIPTION
**Pull request by jasontedor**
When a fatal error is thrown on the network layer, such an error never
makes its way to the uncaught exception handler. This prevents the node
from being torn down if an out of memory error or other fatal error is
thrown while handling HTTP or transport traffic. This commit adds logic
to ensure that such errors bubble their way up to the uncaught exception
handler, even though Netty tries really hard to swallow everything.

Relates #19272

### CODE REVIEW

**Review by s1monw — APPROVED** [review]
LGTM

**Comment by jasontedor** [comment]
To test this, start an instance of Elasticsearch with a 256m heap that does not have this patch applied.

```
$ dd if=/dev/zero of=zero bs=1024k count=512
$ curl -XPOST localhost:9200/i/t/1 --data-binary @zero
```

Elasticsearch will not die. Now apply this patch and test again. Elasticsearch will die with:

```
[2016-11-21T16:27:18,509][ERROR][o.e.b.ElasticsearchUncaughtExceptionHandler] [] fatal error in thread [org.elasticsearch.http.netty4.Netty4HttpRequestHandler#exceptionCaught], exiting
java.lang.OutOfMemoryError: Java heap space
	at io.netty.buffer.PoolArena$HeapArena.newUnpooledChunk(PoolArena.java:661) ~[?:?]
	at io.netty.buffer.PoolArena.allocateHuge(PoolArena.java:246) ~[?:?]
	at io.netty.buffer.PoolArena.allocate(PoolArena.java:224) ~[?:?]
	at io.netty.buffer.PoolArena.allocate(PoolArena.java:141) ~[?:?]
	at io.netty.buffer.PooledByteBufAllocator.newHeapBuffer(PooledByteBufAllocator.java:247) ~[?:?]
	at io.netty.buffer.AbstractByteBufAllocator.heapBuffer(AbstractByteBufAllocator.java:160) ~[?:?]
	at io.netty.buffer.AbstractByteBufAllocator.heapBuffer(AbstractByteBufAllocator.java:151) ~[?:?]
	at io.netty.buffer.CompositeByteBuf.allocBuffer(CompositeByteBuf.java:1653) ~[?:?]
	at io.netty.buffer.CompositeByteBuf.consolidateIfNeeded(CompositeByteBuf.java:405) ~[?:?]
	at io.netty.buffer.CompositeByteBuf.addComponent(CompositeByteBuf.java:196) ~[?:?]
	at io.netty.handler.codec.MessageAggregator.appendPartialContent(MessageAggregator.java:317) ~[?:?]
	at io.netty.handler.codec.MessageAggregator.decode(MessageAggregator.java:282) ~[?:?]
	at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:88) ~[?:?]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:373) ~[?:?]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:359) ~[?:?]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:351) ~[?:?]
	at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:102) ~[?:?]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:373) ~[?:?]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:359) ~[?:?]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:351) ~[?:?]
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:293) ~[?:?]
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:280) ~[?:?]
	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:396) ~[?:?]
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:248) ~[?:?]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:373) ~[?:?]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:359) ~[?:?]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:351) ~[?:?]
	at io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:86) ~[?:?]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:373) ~[?:?]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:359) ~[?:?]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:351) ~[?:?]
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1334) ~[?:?]
```

File: modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4Utils.java [hunk 1]
```diff
@@ -118,4 +118,18 @@ public static void closeChannels(final Collection<Channel> channels) throws IOEx
...
+        if (cause instanceof Error) {
+            /*
+             * Here be dragons. We want to rethrow this so that it bubbles up to the uncaught exception handler. Yet, Netty wraps too many
+             * invocations of user-code in try/catch blocks that swallow all throwables. This means that a rethrow here will not bubble up
+             * to where we want it to. So, we fork a thread and throw the exception from there where Netty can not get to it. We do not wrap
+             * the exception so as to not lose the original cause during exit, so we give the thread a name based on the previous stack
+             * frame so that at least we know where it came from.
+             */
+            final StackTraceElement previous = Thread.currentThread().getStackTrace()[2];
+            new Thread(() -> { throw (Error)cause; }, previous.getClassName() + "#" + previous.getMethodName()).start();
```
Comment (jaymode) [inline modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4Utils.java]: maybe we can try to log the stacktrace in a try-finally block in hopes that we can get the full stacktrace? In the finally we can throw the Error

**Review by jaymode — APPROVED** [review]
left one comment. Other than that LGTM

**Comment by jasontedor** [comment]
retest this please

**Comment by jasontedor** [comment]
Very nice @s1monw, I pushed 96c4116b67002dc8efba22b26d4d8acaf4fec8ec.

**Comment by jasontedor** [comment]
Thanks @s1monw and @jaymode.

**Pull request merged**